server/shared/src/main/scala/org/http4s/server/middleware/BracketRequestResponse.scala

```diff
@@ -121,21 +121,45 @@ object BracketRequestResponse {
         OptionT(
           acquire(request).flatMap(contextRequest =>
             bracketRoutes(contextRequest)
-              .foldF(release(contextRequest.context, None, Outcome.succeeded(F.unit)) *> F.pure(
-                None: Option[Response[F]]))(contextResponse =>
-                F.pure(Some(contextResponse.response.pipeBodyThrough(
-                  _.onFinalizeCaseWeak(ec =>
-                    release(contextRequest.context, Some(contextResponse.context), ec.toOutcome)))))
-              )
+              .foldF(
+                release(contextRequest.context, None, Outcome.succeeded(F.unit)) *>
+                  F.pure(None: Option[Response[F]])
+              )(prepareResponse(contextRequest, _)(release))
               .guaranteeCase {
                   case Outcome.Succeeded(_) =>
                     F.unit
+
+                  // release has been attempted; it's supposed to be run only once
+                  case Outcome.Errored(ReleaseError(_)) =>
+                    F.unit
+
                   case otherwise =>
                     release(contextRequest.context, None, otherwise.void)
-
               })
         ))
-    // format: on
+
+  private def prepareResponse[F[_], A, B](
+      contextRequest: ContextRequest[F, A],
+      contextResponse: ContextResponse[F, B],
+  )(release: (A, Option[B], Outcome[F, Throwable, Unit]) => F[Unit])(
+      implicit F: MonadCancelThrow[F]): F[Option[Response[F]]] =
+    contextResponse.response.entity match {
+      case Entity.Empty | Entity.Strict(_) =>
+        release(contextRequest.context, Option(contextResponse.context), Outcome.succeeded(F.unit))
+          .as(contextResponse.response.some)
+          .handleErrorWith(err => F.raiseError(ReleaseError(err)))
```

Comment:
_Nota bene_: since the response body has been materialized for `Empty`/`Strict` Entities, we are running the release eagerly. And that obviously differs from the behavior of streamed `Entity`, for which release will be logically delayed until `Stream` is materialized. Hence why we have to prevent attempting release twice for non-streamed Entities.